Step3: 자동차 경주 미션

Dev-RubinJo opened the pull request:
Step3 자동차 경주 미션 구현을 완료했습니다.

이번에 객체의 역할에 맞도록 입출력 로직을 담당하는 객체를 따로 만들어서 적용했습니다.

Dev-RubinJo committed:
feat(step3): 자동차 경주 미션

자동차 경주 코드를 작성했습니다.

각각의 로직에 단위테스트를 구현해봤습니다.
필요한 부분에 대하여 에러를 정의하여 처리해뒀습니다.

Dev-RubinJo committed:
Merge remote-tracking branch 'upstream/dev-rubinjo' into step3

# Conflicts:
#	RacingCar.xcodeproj/project.pbxproj

Dev-RubinJo committed:
feat(step3): input, output담당 struct 생성

기존 Racing과 Car에 있던 input, output 메서드를 분리하기 위해 담당 struct를 만들어서 적용했습니다.

wonhee009 commented on RacingCar/RacingCarStep3/Racing.swift:
```diff
@@ -0,0 +1,56 @@
+        try racing()
+    }
+    
+    func createCars(input: Int) {
+        var count = input
+        while count != 0 {
+            cars.append(Car())
+            count -= 1
+        }
+    }
+    
+    func inputTryCount(input: Int) {
+        let count = input
+        tryCount = count
+    }
```
`private`여야할거 같은데요. 혹시 테스트 때문에 열려 있는거라면
- 해당 구조가 올바른지 (하나의 역할만 하고 있는지, 이 역할이 해당 class가 갖는게 맞는지)
- private 메서드를 테스트하는게 맞는지
생각해 보면 좋을거 같습니다 :)

Dev-RubinJo replied:
Input, output 로직을 제외한 모든 로직에 단위테스트를 진행해야한다는 요구사항을 보고 private을 제거했던 것 같습니다. 이야기해주신 내용들을 고려해보고 한번 적용해보도록 하겠습니다.

wonhee009 commented on RacingCar/RacingCarStep3/Racing.swift:
```diff
@@ -0,0 +1,56 @@
+        }
+    }
+    
+    func inputTryCount(input: Int) {
+        let count = input
+        tryCount = count
+    }
```
해당 메서드는 `tryCount`에 input을 지정해주는 역할만 하는거 같은데요.
해당 메서드가 필요한 이유가 있을까요?

```swift
if let count = try? Step3InputOutput.inputTryCount() {
   tryCount = count
}
```
이렇게도 대체 가능할거 같아서용

Dev-RubinJo replied:
위와 같은 형태로 하기엔 createCars도 마찬가지의 형태로 작성할 수 있다고 생각했습니다. 그렇게 되면 car와 tryCount 두가지를 지정해주는 메서드가 생긴다고 파악하여 둘을 각각 메서드로 분리해뒀던것입니다. 혹시 이렇게까지 하는것보다 하나의 메서드 안에서 둘 다 지정해주는게 더 나을 수 있을까요?

Dev-RubinJo replied:
다시한번 코드를 보며 곰곰이 생각을 해보니 inputTryCount메서드는 불필요하게 따로 만들어진 메서드인 것 같습니다.  
```(swift)
tryCount = count
``` 
이 한줄을 수행하기 위해 메서드로 빠져있는 것이 함수 호출시간의 오버헤드가 더 발생하는 것 같군요. 피드백 주신대로 수정을 해보았습니다!

wonhee009 commented on RacingCar/RacingCarStep3/RacingCarStep3Error.swift:
```diff
@@ -0,0 +1,23 @@
+import Foundation
+
+enum RacingCarStep3Error: Error {
+    case inputError
```
사용하는 곳에서는 `RacingCarStep3Error.inputError`로 사용될텐데요.
`RacingCarStep3Error.input`만으로도 에러임을 충분히 인지할 수 있을거 같습니다:)

Dev-RubinJo replied:
엇 그렇군요 Error 접미사를 빼도 명확하겠네요. 감사합니다!

wonhee009 commented on RacingCar/RacingCarStep3/RacingCarStep3Error.swift:
```diff
@@ -0,0 +1,23 @@
+
+import Foundation
+
+enum RacingCarStep3Error: Error {
```
`Step3`라는 네이밍은 없어도 되지 않을까요?

Dev-RubinJo replied:
추후 진행할 Step4, Step5와 구분하기 위해 이렇게 작성해둔 것인데 Step 3, 4, 5와 통합할지 한번 고려해보겠습니다.

wonhee009 commented on RacingCar/RacingCarStep3/Step3Input.swift:
```diff
@@ -0,0 +1,41 @@
+//
+//  Step3Input.swift
```
파일명과 맞춰주세요 :)

Dev-RubinJo replied:
알겠습니다~

wonhee009 commented on RacingCar/main.swift:
```diff
@@ -5,4 +5,3 @@
 // 
 
 import Foundation
-
```
Racing start를 작성해야하지 않을까요?.?

Dev-RubinJo replied:
테스트 위주의 미션이라 main에는 잘 돌아가는지만 확인하고 지워뒀던것인데 남겨두는게 좋을까요?

wonhee009 replied:
아하! 그래도 남겨두는게 어떨지 의견 드려봅니다! (이 부분은 의견여서 편하신대로 해주세요)

Dev-RubinJo replied:
확실히 테스트가 아니라 main에서 작동하는 코드를 남겨두는 것도 괜찮은 것 같아 남겨뒀습니다!

wonhee009 commented on RacingCar/RacingCarStep3/Car.swift:
```diff
@@ -0,0 +1,31 @@
+    
+    @discardableResult
+    func move(randomNumber: Int) throws -> Bool {
+        guard (0 ... 9).contains(randomNumber) else {
+            throw RacingCarStep3Error.invalidNumber
+        }
+        if randomNumber >= 4 {
```
이런 부분들은 validation으로 분리할 수 있을거 같습니다 :)

Dev-RubinJo replied:
덕분에 validation 분리도 따로 나눠야하겠다는 생각을 할 수 있게됐습니다. 감사합니다!

wonhee009 commented on RacingCar/RacingCarStep3/Car.swift:
```diff
@@ -0,0 +1,31 @@
+    }
+    
+    @discardableResult
+    func move(randomNumber: Int) throws -> Bool {
```
Boolean 타입을 반환하는 이유가 있을까요?.?
Racing이 실행되는 부분에서 해당 타입을 사용하는 부분이 없는거 같아서요.

Dev-RubinJo replied:
단순히 테스트를 위해서 넣었던 것인데 에러 반환 유무로 판단을 해야겠네요!

wonhee009 commented on RacingCar/RacingCarStep3/Racing.swift:
```diff
@@ -0,0 +1,56 @@
+    
+    // MARK: - Method
+    func start() throws {
+        if let count = try? Step3InputOutput.inputCarCount() {
+            createCars(input: count)
+        }
+        
+        if let count = try? Step3InputOutput.inputTryCount() {
+            inputTryCount(input: count)
+        }
```
count가 0일 경우나 음수일 경우는 확인 안해도 될까요?

Dev-RubinJo replied:
아 제가 그 부분을 생각을 못했네요. 감사합니다 해당 부분에 대한 처리를 할 수 있도록 하겠습니다.

wonhee009 reviewed (changes requested):
수고하셨습니다 🎉 
몇몇 코멘트 남겨두었으니 확인 부탁드려요 🙏

Dev-RubinJo committed:
refactor(step3): 자동차 경주 미션 피드백 적용

1. 보다 역할에 맞도록 클래스와 메서드를 분리시킴
2. 네이밍에서 중복되는 의미 혹은 불필요한 단어가 포함된 경우에 대한 처리
3. main.swift에 레이싱 경기 시작 로직 적용
4. 메서드에서 불필요한 반환값 제거
5. 입력값 유효성 확인 로직 추가

Dev-RubinJo committed:
fix(step3): modify userInput validation

Validator의 기존 checkValidation(number:)로 모든 곳에 사용하고 있던 것을 분리했습니다.

Dev-RubinJo committed:
fix(step3): separate Racing class with RacingCarController

Racing 클래스에서 차량을 만들어주고, 시행횟수를 지정하는것이 역할이 너무 많아진다는 느낌을 받았습니다.
따라서 Racing에서의 많은 역할을 어떻게 분리하지 생각을 하다가 RacingCarController를 만들어서 Controller에 역할을 분리해뒀습니다.

wonhee009 commented on RacingCar/Car.swift:
```diff
@@ -0,0 +1,25 @@
+    }
+    
+    func move(randomNumber: Int) throws {
+        if try Validator.checkValidation(number: randomNumber) >= 4 {
+            moveCount += 1
+        }
```
4나 1와 같은 상수는 따로 관리하는건 어떤가요?
```swift
final class Car {
   private let canMoveMinNumber = 4
   private let moveCount = 1
}
```

Dev-RubinJo replied:
추후 스펙 변경에 따라 유지보수에 용이할 수 있을만한 방법이군요! 적용해보겠습니다~

wonhee009 commented on RacingCar/Racing.swift:
```diff
@@ -0,0 +1,23 @@
+final class Racing {
+    
+    // MARK: - Method
+    func runRacing(tryCount: Int, cars: [Car], completion: ([Car]) -> Void) throws {
```
`run()`여도 충분히 의미 전달이 가능할거 같습니다:)

Dev-RubinJo replied:
생각해보니 그렇군요! 한번 적용해보겠습니다!

wonhee009 commented on RacingCar/RacingCarController.swift:
```diff
@@ -0,0 +1,32 @@
+
+import Foundation
+
+final class RacingCarController {
```
Car를 생성하고, Racing을 시작하는 역할을 가진 클래스로 이해했습니다.
`RacingController`라는 이름이 좀 더 맞지 않을까 생각이 들었는데 확인 부탁드립니다 :)

현재는 `RacingCarController`에서 'Car를 생성 -> Racing에서 경주 시작'으로 이해했습니다.
`Racing`에서는 경주의 한 사이클만 담당하도록 하고, RacingCarController에서 전반적인 경기를 진행하는건 어떨까 의견 드립니다.
```swift
final class RacingController {
   func start(carCount: Int, racingCount: Int) {
      createCars()
      while racingCount {
         Racing.run(cars)
      }
   }
}
```

Dev-RubinJo replied:
해당 미션이 RacingCar였기에 위와 같이 지었던 것인데 결국 본질을 생각해보면 RacingController가 더 적절한 이름인 것 같군요. 
피드백 해주신 것을 계속 생각하며 Racing의 역할에 대해 생각을 해봤습니다. 레이싱에서 racingCount를 알 필요도 없고 단지 자동차들의 경주만 잘 돌아가게 해주면 된다고 판단을 했습니다. 감사합니다 덕분에 다시 생각해볼 기회를 가졌습니다.

wonhee009 reviewed (approved):
수고하셨습니다 🎉 
코멘트 추가해두었으니 확인 부탁드립니다 🙏 
다음 스텝에 추가해주세요 😄
